mm2src/coins/eth/eth_withdraw.rs

```diff
@@ -0,0 +1,448 @@
+use super::{checksum_address, get_addr_nonce, get_eth_gas_details, u256_to_big_decimal, wei_from_big_decimal,
+            EthCoinType, EthDerivationMethod, EthPrivKeyPolicy, Public, WithdrawError, WithdrawRequest,
+            WithdrawResult, ERC20_CONTRACT, H160, H256};
+use crate::eth::{Action, Address, EthTxFeeDetails, KeyPair, SignedEthTx, UnSignedEthTx};
+use crate::hd_wallet::{HDCoinWithdrawOps, HDWalletOps, WithdrawSenderAddress};
+use crate::rpc_command::init_withdraw::{WithdrawInProgressStatus, WithdrawTaskHandle};
+use crate::{BytesJson, EthCoin, GetWithdrawSenderAddress, TransactionDetails};
+use crate::{CoinWithDerivationMethod, PrivKeyPolicy};
+use async_trait::async_trait;
+use bip32::DerivationPath;
+use common::custom_futures::timeout::FutureTimerExt;
+use common::now_sec;
+use crypto::{CryptoCtx, HwRpcError};
+use ethabi::Token;
+use futures::compat::Future01CompatExt;
+use mm2_core::mm_ctx::MmArc;
+use mm2_err_handle::map_mm_error::MapMmError;
+use mm2_err_handle::mm_error::MmResult;
+use mm2_err_handle::prelude::{MapToMmResult, MmError, OrMmError};
+use std::ops::Deref;
+
+use web3::types::TransactionRequest;
+
+#[async_trait]
+pub trait EthWithdraw
+where
+    Self: Sized + Sync,
+{
+    fn coin(&self) -> &EthCoin;
+
+    fn request(&self) -> &WithdrawRequest;
+
+    #[allow(clippy::result_large_err)]
+    fn on_generating_transaction(&self) -> Result<(), MmError<WithdrawError>>;
+
+    #[allow(clippy::result_large_err)]
+    fn on_finishing(&self) -> Result<(), MmError<WithdrawError>>;
+
+    async fn sign_tx_with_trezor(
+        &self,
+        derivation_path: &DerivationPath,
+        unsigned_tx: &UnSignedEthTx,
+    ) -> Result<SignedEthTx, MmError<WithdrawError>>;
+
+    async fn get_from_address(&self, req: &WithdrawRequest) -> Result<H160, MmError<WithdrawError>> {
+        let coin = self.coin();
+        match req.from {
+            Some(_) => Ok(coin.get_withdraw_sender_address(req).await?.address),
+            None => Ok(coin.derivation_method.single_addr_or_err().await?),
+        }
+    }
+
+    #[allow(clippy::result_large_err)]
+    fn get_key_pair(&self, req: &WithdrawRequest) -> Result<KeyPair, MmError<WithdrawError>> {
+        let coin = self.coin();
+        match req.from {
+            Some(ref from) => {
+                let path_to_coin = &coin
+                    .deref()
+                    .derivation_method
+                    .hd_wallet()
+                    .ok_or(WithdrawError::UnexpectedDerivationMethod)?
+                    .derivation_path;
+                let path_to_address = from.to_address_path(path_to_coin.coin_type())?;
+                let derivation_path = path_to_address.to_derivation_path(path_to_coin)?;
```

Comment:
This code is repeated in `get_from_derivation_path` function we can probably separate it into its own function `get_withdraw_derivation_path`. Then this function `get_key_pair` can be refactored as such
```rust
fn get_key_pair(&self, req: &WithdrawRequest) -> Result<KeyPair, MmError<WithdrawError>> {
    let coin = self.coin();

    if coin.priv_key_policy.is_trezor() {
        return MmError::err(WithdrawError::InternalError("no keypair for hw wallet".to_owned()));
    }

    match &req.from {
        Some(from) => {
            let derivation_path = coin.get_withdraw_derivation_path(from)?;
            let raw_priv_key = coin
                .priv_key_policy
                .hd_wallet_derived_priv_key_or_err(&derivation_path)?;
            KeyPair::from_secret_slice(raw_priv_key.as_slice()).map_to_mm(|e| WithdrawError::InternalError(e.to_string()))
        },
        None => coin.priv_key_policy.activated_key_or_err(),
    }
}
```

Reply:
fixed in [bfe7a29](https://github.com/KomodoPlatform/komodo-defi-framework/pull/2005/commits/bfe7a29f7bf4a4c280b29f393800d614a2077c72) (note I used the other way around fn names that you suggested)